h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -1333,8 +1333,10 @@ else if (_parms._weights_column != null && _weights != null && !_weights.isBinar
       if (restructured)
         _train.restructure(_train.names(), vecs);
     }
-    assert (!expensive || _valid==null || Arrays.equals(_train._names, _valid._names) || _parms._categorical_encoding == Model.Parameters.CategoricalEncodingScheme.Binary);
-    if (_valid!=null && !Arrays.equals(_train._names, _valid._names) && _parms._categorical_encoding == Model.Parameters.CategoricalEncodingScheme.Binary) {
+    boolean names_may_differ = _parms._categorical_encoding == Model.Parameters.CategoricalEncodingScheme.Binary;
+    boolean names_differ = _valid !=null && !Arrays.equals(_train._names, _valid._names);
+    assert (!expensive || names_may_differ || !names_differ);
```

Comment:
this currently fails if I add EnumLimited to the `checkCategoricals` in adaptTestToTrain due to double encoding (enum limited encoding is not idempotent...)